node-wasm/src/worker/channel.rs

```diff
@@ -0,0 +1,191 @@
+pub(super) struct ClientId(usize);
+
+impl fmt::Display for ClientId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Client({})", self.0)
+    }
+}
+
+pub(super) enum WorkerMessage {
+    NewConnection(MessagePort),
+    Command((NodeCommand, ClientId)),
+}
+
+pub(super) struct WorkerMessageServer {
+    // same onconnect callback is used throughtout entire Worker lifetime.
+    // Keep a reference to make sure it doesn't get dropped.
+    _onconnect_callback: Closure<dyn Fn(MessageEvent)>,
+
+    // keep a MessagePort for each client to send messages over, as well as callback responsible
+    // for forwarding messages back
+    clients: Vec<WorkerClientConnection>,
+
+    // sends events back to the main loop for processing
+    command_channel: mpsc::Sender<WorkerMessage>,
+}
+
+impl WorkerMessageServer {
+    pub fn new(command_channel: mpsc::Sender<WorkerMessage>) -> Self {
+        let near_tx = command_channel.clone();
+        let onconnect_callback: Closure<dyn Fn(MessageEvent)> =
+            Closure::new(move |ev: MessageEvent| {
+                let local_tx = near_tx.clone();
+                spawn_local(async move {
+                    let Ok(port) = ev.ports().at(0).dyn_into() else {
+                        error!("received connection event without MessagePort, should not happen");
+                        return;
+                    };
+
+                    if let Err(e) = local_tx.send(WorkerMessage::NewConnection(port)).await {
+                        error!("command channel inside worker closed, should not happen: {e}");
+                    }
+                })
+            });
+
+        let worker_scope = SharedWorker::worker_self();
+        worker_scope.set_onconnect(Some(onconnect_callback.as_ref().unchecked_ref()));
+
+        Self {
+            _onconnect_callback: onconnect_callback,
+            clients: Vec::with_capacity(1), // we usually expect to have exactly one client
+            command_channel,
+        }
+    }
+
+    pub fn respond_to(&self, client: ClientId, msg: WorkerResponse) {
+        self.send_response(client, Ok(msg))
+    }
+
+    pub fn respond_err_to(&self, client: ClientId, error: WorkerError) {
+        self.send_response(client, Err(error))
+    }
+
+    pub fn add(&mut self, port: MessagePort) {
+        let client_id = self.clients.len();
+
+        let near_tx = self.command_channel.clone();
+        let client_message_callback: Closure<dyn Fn(MessageEvent)> =
+            Closure::new(move |ev: MessageEvent| {
+                let local_tx = near_tx.clone();
+                spawn_local(async move {
+                    let client_id = ClientId(client_id);
+                    let message_data = ev.data();
+                    let Ok(command) = from_value::<NodeCommand>(message_data) else {
+                        warn!("could not deserialize message from client {client_id}");
+                        return;
+                    };
```

Comment:
We should still respond because client is waiting. Maybe we can do something like `.ok_or(NodeCommand::Invalid)` and then worker respond with something like `WokerError::InvalidCommand`.

Reply:
Good point. I've added new worker message and handling of it at the `run_worker` level.